### PR TITLE
fix: #2204, #2211 resolve loop

### PR DESCRIPTION
**Pull request by technohippy**
⯈fix: #2204 
⯈fix: #2211

- [x] I've read the contributing guidelines
- [x] I've added my name and email to the NOTICE file

### CODE REVIEW

**Review by romdotdog — COMMENTED** [review]
You should use an O(1) `Set<Expression>` instead of an `Expression[]`

File: src/resolver.ts [hunk 1]
```diff
@@ -1006,6 +1009,21 @@ export class Resolver extends DiagnosticEmitter {
     ctxType: Type = Type.auto,
     /** How to proceed with eventual diagnostics. */
     reportMode: ReportMode = ReportMode.REPORT
+  ): Type | null {
+    if (this.resolvingExpressions.has(node)) return null;
+
+    this.resolvingExpressions.add(node);
+    const resolved = this.unsafeResolveExpression(node, ctxFlow, ctxType, reportMode);
+    this.resolvingExpressions.delete(node);
+    return resolved;
```
Comment (dcodeIO) [inline src/resolver.ts:1018]: Can cache the property access in a local here:
```suggestion
    const resolvingExpressions = this.resolvingExpressions;
    if (resolvingExpressions.has(node)) return null;
    resolvingExpressions.add(node);
    const resolved = this.unsafeResolveExpression(node, ctxFlow, ctxType, reportMode);
    resolvingExpressions.delete(node);
    return resolved;
```

File: src/resolver.ts [hunk 2]
```diff
@@ -1006,6 +1009,21 @@ export class Resolver extends DiagnosticEmitter {
...
+  }
+
+  /** Resolves an expression to its static type. (may cause stack overflow) */
+  private unsafeResolveExpression(
```
Comment (dcodeIO) [inline src/resolver.ts:1022]: Perhaps...?
```suggestion
  private doResolveExpression(
```
I think the Tokenizer uses a similar `unsafe` prefix for a method, but it's not *technically* unsafe.

**Comment by dcodeIO** [comment]
Thanks! :)

**Pull request merged**